File: cpp/house_robber.cpp
// House Robber.
//
// Cannot rob two adjacent houses. Maximize total robbed.
//
// Time: O(n)
// Space: O(1)

#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

int rob(const std::vector<int>& nums) {
    int prev_one = 0;
    int prev_two = 0;
    for (int x : nums) {
        int new_one = std::max(prev_one, prev_two + x);
        prev_two = prev_one;
        prev_one = new_one;
    }
    return prev_one;
}

int main() {
    assert(rob({1, 2, 3, 1}) == 4);
    assert(rob({2, 7, 9, 3, 1}) == 12);
    assert(rob({}) == 0);
    std::cout << "OK\n";
    return 0;
}

File: java/HouseRobber.java
/*
 * House Robber.
 *
 * Cannot rob two adjacent houses. Maximize total robbed.
 *
 * Time: O(n)
 * Space: O(1)
 */

public class HouseRobber {

    public static int rob(int[] nums) {
        int prevOne = 0;
        int prevTwo = 0;
        for (int x : nums) {
            int newOne = Math.max(prevOne, prevTwo + x);
            prevTwo = prevOne;
            prevOne = newOne;
        }
        return prevOne;
    }

    public static void main(String[] args) {
        assert rob(new int[] {1, 2, 3, 1}) == 4;
        assert rob(new int[] {2, 7, 9, 3, 1}) == 12;
        assert rob(new int[] {}) == 0;
        System.out.println("OK");
    }
}
